resolver_wrapper.go

```diff
@@ -78,7 +79,11 @@ func (ccr *ccResolverWrapper) start() error {
 			Authority:            ccr.cc.authority,
 		}
 		var err error
-		ccr.resolver, err = ccr.cc.resolverBuilder.Build(ccr.cc.parsedTarget, ccr, opts)
+		if ccr.cc.dopts.copts.Dialer == nil && (ccr.cc.dopts.copts.UseProxy && !ccr.cc.dopts.copts.TargetResolutionEnabled) {
+			ccr.resolver, err = delegatingresolver.New(ccr.cc.parsedTarget, ccr, opts, ccr.cc.resolverBuilder)
+		} else {
+			ccr.resolver, err = ccr.cc.resolverBuilder.Build(ccr.cc.parsedTarget, ccr, opts)
+		}
```

Comment:
I don't remember if I asked you to invert the conditional in a previous comment, but this feels quite hard to read.

How about:
```
// The delegating resolver is not used if WithNoProxy or WithCustomDialer is set. 
// It is not used when explicitly disabled with TargetResolutionEnabled as well. 
// In these cases, the normal name resolver determined by the scheme will be used directly.
if ccr.cc.dopts.copts.Dialer != nil || ccr.cc.dopts.copts.UseProxy || ccr.cc.dopts.copts.TargetResolutionEnabled) {
	ccr.resolver, err = ccr.cc.resolverBuilder.Build(ccr.cc.parsedTarget, ccr, opts)
} else {
	ccr.resolver, err = delegatingresolver.New(ccr.cc.parsedTarget, ccr, opts, ccr.cc.resolverBuilder)
}
```